### PR TITLE
As type supports sequence of arrays from multiple devices

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
Added support for `asarray` to create arrays from sequences of usm_ndarray objects allocated on different devices, and other Python sequences or numpy arrays.

```
In [1]: import dpctl, dpctl.tensor as dpt

In [2]: m = dpt.zeros((2, 4), dtype="i2", device="cpu")

In [3]: w = dpt.full(4, -1, device="opencl:gpu")

In [4]: res = dpt.asarray([m, [w, [0,] * 4 ]], dtype="f4", device="cpu")

In [5]: res
Out[5]:
usm_ndarray([[[ 0.,  0.,  0.,  0.],
              [ 0.,  0.,  0.,  0.]],

             [[-1., -1., -1., -1.],
              [ 0.,  0.,  0.,  0.]]], dtype=float32)

In [6]: res.device
Out[6]: Device(opencl:cpu:0)
```

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you opening the PR as a draft?

### CODE REVIEW

File: dpctl/tensor/_ctors.py
Comment (antonwolfy) [inline dpctl/tensor/_ctors.py:1]: The code below fails:
```
# import dpnp, numpy, dpctl, dpctl.tensor as dpt
# w = dpt.full(4, -1, device="opencl:gpu")
# res = dpt.asarray([w, range(4)], dtype="f4")

TypeError                                 Traceback (most recent call last)
Cell In[6], line 1
----> 1 res = dpt.asarray([w, range(4)], dtype="f4")

File /localdisk/work/antonvol/soft/miniconda3/envs/dpnp_py39_ext/lib/python3.9/site-packages/dpctl/tensor/_ctors.py:646, in asarray(obj, dtype, device, copy, usm_type, sycl_queue, order)
    644 if len(devs) == 1:
    645     seq_dev = devs[0]
--> 646     return _asarray_from_seq_single_device(
    647         obj,
    648         seq_shape,
    649         seq_dt,
    650         seq_dev,
    651         dtype=dtype,
    652         usm_type=usm_type,
    653         sycl_queue=sycl_queue,
    654         order=order,
    655     )
    656 else:
    657     if sycl_queue is None:

File /localdisk/work/antonvol/soft/miniconda3/envs/dpnp_py39_ext/lib/python3.9/site-packages/dpctl/tensor/_ctors.py:480, in _asarray_from_seq_single_device(obj, seq_shape, seq_dt, seq_dev, dtype, usm_type, sycl_queue, order)
    473     exec_q = dpctl.utils.get_execution_queue(
    474         (
    475             sycl_queue,
    476             seq_dev,
    477         )
    478     )
    479     alloc_q = sycl_queue
--> 480 return _asarray_from_seq(
    481     obj,
    482     seq_shape,
    483     seq_dt,
    484     alloc_q,
    485     exec_q,
    486     dtype=dtype,
    487     usm_type=usm_type,
    488     order=order,
    489 )

File /localdisk/work/antonvol/soft/miniconda3/envs/dpnp_py39_ext/lib/python3.9/site-packages/dpctl/tensor/_ctors.py:444, in _asarray_from_seq(seq_obj, seq_shape, seq_dt, alloc_q, exec_q, dtype, usm_type, order)
    436 res = dpt.empty(
    437     seq_shape,
    438     dtype=dtype,
   (...)
    441     order=order,
    442 )
    443 ht_events = []
--> 444 _device_copy_walker(seq_obj, res, ht_events)
    445 dpctl.SyclEvent.wait_for(ht_events)
    446 return res

File /localdisk/work/antonvol/soft/miniconda3/envs/dpnp_py39_ext/lib/python3.9/site-packages/dpctl/tensor/_ctors.py:361, in _device_copy_walker(seq_o, res, events)
    359 if isinstance(seq_o, (list, tuple)):
    360     for i, el in enumerate(seq_o):
--> 361         _device_copy_walker(el, res[i], events)
    362     return
    363 raise TypeError

File /localdisk/work/antonvol/soft/miniconda3/envs/dpnp_py39_ext/lib/python3.9/site-packages/dpctl/tensor/_ctors.py:363, in _device_copy_walker(seq_o, res, events)
    361         _device_copy_walker(el, res[i], events)
    362     return
--> 363 raise TypeError

TypeError:
```

Comment (oleksandr-pavlyk) [reply dpctl/tensor/_ctors.py:1]: Thank you for the report @antonwolfy 

This issue should be fixed now. Please take another look.

File: dpctl/tensor/_ctors.py [hunk 1]
```diff
@@ -576,16 +628,51 @@ def asarray(
...
+                    dtype=dtype,
+                    usm_type=usm_type,
+                    sycl_queue=sycl_queue,
+                    order=order,
+                )
+            else:
+                if sycl_queue is None:
+                    raise dpctl.utils.ExecutionPlacementError(
+                        "Please specify `device` or `sycl_queue` keyword "
+                        "argument to determine where to allocated the "
```
[the remark this answers is not in the document]
Comment (oleksandr-pavlyk) [reply dpctl/tensor/_ctors.py]: Corrected.

File: dpctl/tensor/_ctors.py [hunk 2]
```diff
@@ -329,6 +329,10 @@ def _usm_types_walker(o, usm_types_list):
         usm_ar = _usm_ndarray_from_suai(o)
         usm_types_list.append(usm_ar.usm_type)
         return
+    if _is_object_with_buffer_protocol(o):
+        return
+    if isinstance(o, (int, bool, float, complex)):
+        return
     if isinstance(o, (list, tuple)):
```
[the remark this answers is not in the document]
Comment (oleksandr-pavlyk) [reply dpctl/tensor/_ctors.py]: This example should be working now.

**Review by antonwolfy — APPROVED** [review]

**Pull request merged**